# DCP Data Consumer Experience


## Overview
Welcome to the DCP Vignette repository, containing walkthrough tutorials to help you get started with the DCP primarily via command-line access. For downstream application development, please refer to the [HCA DCP API documentation](https://prod.data.humancellatlas.org/apis)  Before the DCP exits beta, the vignettes below are all considered experimental.

* ***DCP-supported Vignettes*** (work with the current version of the DCP)
* ***Contributed Vignettes*** (generated by DCP and external contributors, but are not validated before each release of the DCP)
* ***Beta Vignettes*** These are either in-progress, or determined to have fallen behind the current version of the software. If you find a Contributed Vignette that doesn't work as it should, it should (1) be moved to the Beta Vignettes section and the author notified, *or* (2) a revision made with a new pull request to the repository.
* ***Proposed Vignettes*** These are tutorials proposed but do not yet have a shepherd.

The DCP welcomes any contributed notebooks or other tutorials to the list below. You can create your own branch and submit a pull request. 

## DCP-supported Vignettes
None currently

## Contributed Vignettes
| Status | Category | Task |
| --- | --- | --- |
| :white_check_mark: | Data Visualization | [Use Scanpy to Visualize Matrix data using tSNE and UMAP](nov2018_demo/hca_demo_scanpy.ipynb)
| :white_check_mark: | Fundamentals | [Install the HCA CLI](tasks/Install)

## Beta Vignettes

| Status | Category | Task |
| --- | --- | --- |
| :large_orange_diamond: | Fundamentals | [Log in to the DSS](tasks/Log%20In) |
| :full_moon: | Primary and Secondary Data | [Stream FASTQs and Calculate k-mers with SPARK](tasks/SPARK) |
| :full_moon: | Raw Data Exploration | [Download any BAM file](tasks/Download%20BAM) |
| :large_orange_diamond: | Fundamentals | Download FASTQs associated with a known sample ID |
| :large_orange_diamond: | Secondary Data | [Download SmartSeq2 expression matrix as an input to scanpy](tasks/Download%20Expression%20Matrix%20for%20Scanpy) |
| :full_moon: | Search | [Find out how many liver (whatever) cells are available](tasks/Find%20Cell%20Type%20Count) |
| :full_moon: | Raw Data Exploration | [Download all bundles for T-cells sequenced with 10x](tasks/Download%2010x%20Seq%20T-cell%20Bundles) |



## Proposed Vignettes
| Status | Category | Task |
| --- | --- | --- |
| :large_orange_diamond: | Compare QC metrics between one experiment and another |
| :white_circle: | Check gene expression of key markers for immune cells |
| :white_circle: | Compare averaged gene expression of neurons to bulk controls |
| :white_circle: | Find most variable genes in monocytes across experiments |
| :white_circle: | Identify most differentially expressed genes between t-cells and monocytes |
| :white_circle: | Discover novel liver cell types by cross-referencing HCA single cells with known datasets |

##### Legend - User Experience
| :white_check_mark: | :full_moon: | :large_orange_diamond: | :white_circle: |
| :---: | :---: | :---: | :---: |
|Works great! | A little kludgy | Requires workarounds | Proposed |
